rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,258 @@
+
+Many new engineers can find it hard to adapt to a new codebase, especially when the coding patterns and practices are fairly new and in a constant state of flux
+
+## Problem statement
+
+<!--
+Why are we making this change? What problem are we solving? What do we expect to gain from this?
+
+This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.
+
+In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
+-->
+
+Many new developers might have questions on the patterns that we follow to write a React component. This RFC attempts to commit and possibly improve our code patterns in writing so that there might be a reference for new (and old) members of the team.
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### Sample.tsx
+
+The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component
+
+```typescript
+export const Sample = React.forwardRef<HTMLElement, SampleProps>((props, ref) => {
+  const state = useSample(props, ref);
+  useSampleStyles(state);
+
+  return renderSample(state);
+});
+```
+
+#### useSample.ts
+
+Accepts the component props and handles the internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props, we set defaults internally but if consumer wants component variants with different defaults, they can set them
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps): SampleState => {
+  const resolvedRef = ref || React.useRef();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
```

Comment:
```suggestion
  const draftState = mergeProps(
```

Let's try to align naming

Reply:
Let's have a talk about this offline

Reply:
So turns out `mergeProps` is actually supposed return state, this needs to be reflected

Reply:
Yes I feel similarly to @Hotell that it's not very intuitive that `mergeProps` actually returns the first draft of `state`

I've renamed all instances to be `state`